File: api-gateway/internal/middleware/metrics.go
package middleware

import (
	"net/http"
	"strconv"
	"time"

	"github.com/go-chi/chi/v5"
	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promauto"
)

var (
	requestsTotal = promauto.NewCounterVec(prometheus.CounterOpts{
		Name: "gateway_http_requests_total",
		Help: "Total HTTP requests handled by the gateway.",
	}, []string{"code", "method", "route"})

	requestDuration = promauto.NewHistogramVec(prometheus.HistogramOpts{
		Name:    "gateway_http_request_duration_seconds",
		Help:    "HTTP request duration in seconds.",
		Buckets: prometheus.DefBuckets,
	}, []string{"code", "method", "route"})
)

// Metrics records RED metrics per request. The route label uses the chi route
// pattern (not the raw path) to keep label cardinality bounded.
func Metrics(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path == "/metrics" || r.URL.Path == "/health" {
			next.ServeHTTP(w, r)
			return
		}

		start := time.Now()
		rec := &responseRecorder{ResponseWriter: w, status: http.StatusOK}

		next.ServeHTTP(rec, r)

		route := chi.RouteContext(r.Context()).RoutePattern()
		if route == "" {
			route = "unmatched"
		}
		code := strconv.Itoa(rec.status)
		requestsTotal.WithLabelValues(code, r.Method, route).Inc()
		requestDuration.WithLabelValues(code, r.Method, route).Observe(time.Since(start).Seconds())
	})
}

File: file-service/src/plugins/metrics.ts
import fp from "fastify-plugin";
import { FastifyInstance } from "fastify";
import client from "prom-client";

/**
 * RED metrics + /metrics exposition for Prometheus (platform-observability).
 * Route label uses the matched route pattern, never the raw URL, to keep
 * label cardinality bounded. /metrics and /health are excluded from stats.
 */
export default fp(
  async (fastify: FastifyInstance) => {
    const registry = new client.Registry();
    client.collectDefaultMetrics({ register: registry });

    const requestsTotal = new client.Counter({
      name: "file_service_http_requests_total",
      help: "Total HTTP requests handled.",
      labelNames: ["code", "method", "route"],
      registers: [registry],
    });

    const requestDuration = new client.Histogram({
      name: "file_service_http_request_duration_seconds",
      help: "HTTP request duration in seconds.",
      labelNames: ["code", "method", "route"],
      registers: [registry],
    });

    fastify.addHook("onResponse", async (request, reply) => {
      const route = request.routeOptions.url ?? "unmatched";
      if (route === "/metrics" || route === "/health") return;

      const labels = {
        code: String(reply.statusCode),
        method: request.method,
        route,
      };
      requestsTotal.inc(labels);
      requestDuration.observe(labels, reply.elapsedTime / 1000);
    });

    fastify.get("/metrics", async (_request, reply) => {
      reply.type(registry.contentType);
      return registry.metrics();
    });
  },
  { name: "metrics" },
);
